File: examples/java-maven/src/main/java/com/example/impl/InMemoryUserRepository.java
package com.example.impl;

import com.example.interfaces.Repository;
import com.example.models.Status;
import com.example.models.User;

import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * In-memory implementation of User repository.
 */
public class InMemoryUserRepository implements Repository<User> {

    private final Map<Long, User> storage = new HashMap<>();
    private final AtomicLong idGenerator = new AtomicLong(1);

    @Override
    public Optional<User> findById(Long id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public List<User> findAll() {
        return new ArrayList<>(storage.values());
    }

    @Override
    public User save(User entity) {
        User toSave = entity.id() == null
                ? entity.withId(idGenerator.getAndIncrement())
                : entity;
        storage.put(toSave.id(), toSave);
        return toSave;
    }

    @Override
    public boolean delete(Long id) {
        return storage.remove(id) != null;
    }

    public Optional<User> findByEmail(String email) {
        return storage.values().stream()
                .filter(u -> u.email().equals(email))
                .findFirst();
    }

    public List<User> findByStatus(Status status) {
        return storage.values().stream()
                .filter(u -> u.status() == status)
                .toList();
    }

    public List<User> findActive() {
        return findByStatus(Status.ACTIVE);
    }

    public List<User> findByNameContaining(String namePart) {
        return storage.values().stream()
                .filter(u -> u.name().toLowerCase().contains(namePart.toLowerCase()))
                .toList();
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/impl/InMemoryUserRepository.kt
package com.example.impl

import com.example.interfaces.Repository
import com.example.models.Status
import com.example.models.User
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * In-memory implementation of User repository.
 */
class InMemoryUserRepository : Repository<User> {

    private val storage = ConcurrentHashMap<Long, User>()
    private val idGenerator = AtomicLong(1)

    override fun findById(id: Long): User? = storage[id]

    override fun findAll(): List<User> = storage.values.toList()

    override fun save(entity: User): User {
        val toSave = if (entity.id == null) {
            entity.withId(idGenerator.getAndIncrement())
        } else {
            entity
        }
        storage[toSave.id!!] = toSave
        return toSave
    }

    override fun delete(id: Long): Boolean = storage.remove(id) != null

    fun findByEmail(email: String): User? =
        storage.values.find { it.email == email }

    fun findByStatus(status: Status): List<User> =
        storage.values.filter { it.status == status }

    fun findActive(): List<User> = findByStatus(Status.ACTIVE)

    fun findByNameContaining(namePart: String): List<User> =
        storage.values.filter { it.name.contains(namePart, ignoreCase = true) }
}
